README.md
# easey

Easey adds easing to Modest Maps, giving it a serious workout.

* [Docs](https://github.com/mapbox/easey/wiki)
* [Online tests](http://mapbox.com/easey/test/)
* [Online demo](http://mapbox.com/easey/)

## Contributors

This software is made awesome by people like you. If you see something
to improve, send a pull request!

* [aibram (Ansis Brammanis)](https://github.com/aibram)
* [mikemorris (Mike Morris)](https://github.com/mikemorris)


CHANGELOG
### 2.0.7

* Touch & zoom will work on platforms that lack `e.scale`, including Chrome
  Desktop

### 2.0.6

* Touch handlers now work on Android

## 2.0.5

* .optimal() properly handlers interrupting running animations

## 2.0.4

* Remove unused onHold function
* Fix touch handler

## 2.0.3

* Dispatches "zoomed" when zoom level changes
* Touch handler now able to pinch more than one zoom level
* Several optimizations

## 2.0.2

* Fix bug breaking `easey.optimal()`

## 2.0.1

* `from` and `to` are now set to the map center by
  default.

## 2.0.0

* `easey` handlers are now moved to `easey_handlers`
  instead of the `easey` function/object.

## 1.0.3

* Adds API.md document

## 1.0.2

* Fix concatenated files by beginning source with `;`
* Fix style issues


test/spec/Easey.js
describe("Easey", function() {
    function Receiver() { }
    Receiver.prototype.receive = function() { };

    var map, sink;

    beforeEach(function() {
        sink = new Receiver();
        var map_div = document.createElement('div');
        map = new MM.Map(map_div, new MM.TemplatedLayer('http://b.tile.openstreetmap.org/{Z}/{X}/{Y}.png'));
    });

    it('assigns the map var correctly', function() {
        var ease = easey();
        ease.map(map);
        expect(ease.map()).toEqual(map);
    });

    it('automatically sets to', function() {
        var ease = easey();
        ease.map(map);
        expect(ease.to()).toEqual(map.coordinate);
    });

    it('zooms the to coordinate with zoom()', function() {
        var ease = easey();
        ease.map(map);
        ease.zoom(10);
        expect(ease.to().zoom).toEqual(10);
    });

    it('correctly interpolates between two coordinates', function() {
        easey().map(map)
            .from(new MM.Coordinate(0, 10, 0))
            .to(new MM.Coordinate(0, 0, 0))
            .easing('linear')
            .t(0.5);

        expect(map.coordinate.column).toEqual(5);
        expect(map.coordinate.row).toEqual(0);
        expect(map.coordinate.zoom).toEqual(0);
    });

    it('predicts the future correctly', function() {
        var ease = easey();
        ease.map(map).from(new MM.Coordinate(0, 10, 0))
        .to(new MM.Coordinate(0, 0, 0));
        var future = ease.future(10);
        expect(future.length).toEqual(10);
        expect(future[0].column).toEqual(10);
        expect(future[9].column).toEqual(0);
    });

    it('moves the map quickly', function() {
        var ease = easey();
        ease.map(map).from(new MM.Coordinate(0, 10, 0))
        .to(new MM.Coordinate(0, 0, 0));
        runs(function() {
            ease.run(10);
            expect(ease.running()).toBeTruthy();
        });
        waits(200);
        runs(function() {
            expect(map.coordinate.column).toEqual(0);
            expect(map.coordinate.row).toEqual(0);
            expect(map.coordinate.zoom).toEqual(0);
            expect(ease.running()).toBeFalsy();
        });
    });

    it('calls a callback after finishing an ease', function() {
        var ease = easey();
        ease.map(map).from(new MM.Coordinate(0, 10, 0))
        .to(new MM.Coordinate(0, 0, 0));
        spyOn(sink, 'receive');
        runs(function() {
            ease.run(10, sink.receive);
            expect(ease.running()).toBeTruthy();
        });
        waits(200);
        runs(function() {
            expect(sink.receive).toHaveBeenCalledWith(map);
            expect(ease.running()).toBeFalsy();
        });
    });

    it('ends up at the correct coord after an optimal zoom/pan', function() {
        map.setSize(new MM.Point(10, 10));
        var ease = easey();
        ease.map(map).from(new MM.Coordinate(2, 2, 2))
            .to(new MM.Coordinate(1, 1, 1));
        runs(function() {
            ease.optimal(20);
            expect(ease.running()).toBeTruthy();
        });
        waits(200);
        runs(function() {
            expect(map.coordinate.column).toEqual(1);
            expect(map.coordinate.row).toEqual(1);
            expect(map.coordinate.zoom).toEqual(1);
            expect(ease.running()).toBeFalsy();
        });
    });

    it('can be stopped', function() {
        var ease = easey();
        spyOn(sink, 'receive');
        runs(function() {
            ease.map(map).from(new MM.Coordinate(0, 10, 0))
                .to(new MM.Coordinate(0, 0, 0)).run(10);
            expect(ease.running()).toBeTruthy();
            ease.stop(sink.receive);
        });
        waits(500);
        runs(function() {
            expect(sink.receive).toHaveBeenCalled();
            expect(ease.running()).toBeFalsy();
        });
    });

    it('resets from after a run', function() {
        var ease = easey();
        ease.map(map).from(new MM.Coordinate(0, 10, 0))
            .to(new MM.Coordinate(0, 0, 0)).run(1);

        waits(50);
        runs(function() {
            expect(ease.from()).toEqual(undefined);
        });
    });
});


time.js
var map;

window.onload = function() {
    var mm = com.modestmaps;
    var dmap = document.getElementById('map');
    wax.tilejson('http://a.tiles.mapbox.com/v3/mapbox.mapbox-streets.jsonp',
        function(tj) {
        map = new com.modestmaps.Map(dmap, new wax.mm.connector(tj), null, []);
        map.setCenterZoom(new com.modestmaps.Location(50, 50), 4);

        var right = document.getElementById('right');

        var positions = [
          map.locationCoordinate({ lat: 0, lon: 0 }).zoomTo(4),
          map.locationCoordinate({ lat: 20, lon: -50 }).zoomTo(4),
          map.locationCoordinate({ lat: 50, lon: 50 }).zoomTo(4)];

        easey.map(map);

        function update() {
          var pos = right.scrollTop / 500;

          easey.from(positions[Math.floor(pos)])
            .to(positions[Math.ceil(pos)])
            .t(pos - Math.floor(pos));
        }

        right.addEventListener('scroll', update, false);
    });
};


braid.js
var map;

window.onload = function() {
    var mm = com.modestmaps;
    var dmap = document.getElementById('map');
    wax.tilejson('http://a.tiles.mapbox.com/v3/mapbox.mapbox-streets.jsonp',
        function(tj) {
        map = new com.modestmaps.Map(dmap, new wax.mm.connector(tj), null, [
          easey.DragHandler(),
          easey.BraidHandler(),
          easey.DoubleClickHandler()
        ]);
        map.setCenterZoom(new com.modestmaps.Location(50, 50), 4);

        var right = document.getElementById('right');

        var positions = [
          map.locationCoordinate({ lat: 0, lon: 0 }).zoomTo(4),
          map.locationCoordinate({ lat: 20, lon: -50 }).zoomTo(4),
          map.locationCoordinate({ lat: 50, lon: 50 }).zoomTo(4)];

        easey().map(map);

        function update() {
          var pos = right.scrollTop / 500;

          easey.from(positions[Math.floor(pos)])
            .to(positions[Math.ceil(pos)])
            .t(pos - Math.floor(pos));
        }

        right.addEventListener('scroll', update, false);
    });
};


index.js
var map;

window.onload = function() {
    var mm = com.modestmaps;
    var dmap = document.getElementById('map');
    wax.tilejson('http://a.tiles.mapbox.com/v3/examples.map-vyofok3q.jsonp',
        function(tj) {
        map = new com.modestmaps.Map(dmap,
            new wax.mm.connector(tj), null, [
                easey_handlers.DragHandler(),
                easey_handlers.TouchHandler(),
                easey_handlers.MouseWheelHandler(),
                easey_handlers.DoubleClickHandler()
            ]);
        map.setCenterZoom(new com.modestmaps.Location(-10, 50), 3);

        map.addCallback('zoomed', function() {
            // console.log(map.getZoom());
        });

        var pres = document.getElementsByTagName('pre');

        for (var i = 0; i < pres.length; i++) {
          pres[i].onclick = function() {
            eval(this.innerHTML);
          };
        }

        var scrolly = document.getElementById('scrolly');

        var positions = [
          map.locationCoordinate({ lat: 33.5, lon: 65.6 }).zoomTo(6),
          map.locationCoordinate({ lat: 33.1, lon: 44.6 }).zoomTo(6),
          map.locationCoordinate({ lat: 28.7, lon: 69.2 }).zoomTo(6)];

        var ea = easey().map(map).easing('easeInOut');

        function update() {
          var pos = scrolly.scrollTop / 200;

          ea.from(positions[Math.floor(pos)])
            .to(positions[Math.ceil(pos)])
            .t(pos - Math.floor(pos));
        }

        scrolly.addEventListener('scroll', update, false);
    });
};


src/easey.js
;(function(context, MM) {
    var easey = function() {
        var easey = {},
            running = false,
            abort = false, // killswitch for transitions
            abortCallback; // callback called when aborted

        var easings = {
            easeIn: function(t) { return t * t; },
            easeOut: function(t) { return Math.sin(t * Math.PI / 2); },
            easeInOut: function(t) { return (1 - Math.cos(Math.PI * t)) / 2; },
            linear: function(t) { return t; }
        };
        var easing = easings.easeOut;

        // to is the singular coordinate that any transition is based off
        // three dimensions:
        //
        // * to
        // * time
        // * path
        var from, to, map;

        easey.stop = function(callback) {
            abort = true;
            from = undefined;
            abortCallback = callback;
        };

        easey.running = function() {
            return running;
        };

        easey.point = function(x) {
            to = map.pointCoordinate(x);
            return easey;
        };

        easey.zoom = function(x) {
            if (!to) to = map.coordinate.copy();
            to = map.enforceZoomLimits(to.zoomTo(x));
            return easey;
        };

        easey.location = function(x) {
            to = map.locationCoordinate(x);
            return easey;
        };

        easey.from = function(x) {
            if (!arguments.length) return from ? from.copy() : from;
            from = x.copy();
            return easey;
        };

        easey.to = function(x) {
            if (!arguments.length) return to.copy();
            to = map.enforceZoomLimits(x.copy());
            return easey;
        };

        easey.path = function(x) {
            path = paths[x];
            return easey;
        };

        easey.easing = function(x) {
            easing = easings[x];
            return easey;
        };

        easey.map = function(x) {
            if (!arguments.length) return map;
            map = x;
            return easey;
        };

        function interp(a, b, p) {
            if (p === 0) return a;
            if (p === 1) return b;
            return a + ((b - a) * p);
        }

        var paths = {},
            static_coord = new MM.Coordinate(0, 0, 0);

        // The screen path simply moves between
        // coordinates in a non-geographical way
        paths.screen = function(a, b, t, static_coord) {
            var zoom_lerp = interp(a.zoom, b.zoom, t);
            if (static_coord) {
                static_coord.row = interp(
                    a.row,
                    b.row * Math.pow(2, a.zoom - b.zoom),
                    t) * Math.pow(2, zoom_lerp - a.zoom);
                static_coord.column = interp(
                    a.column,
                    b.column * Math.pow(2, a.zoom - b.zoom),
                    t) * Math.pow(2, zoom_lerp - a.zoom);
                static_coord.zoom = zoom_lerp;
            } else {
                return new MM.Coordinate(
                    interp(a.row,
                        b.row * Math.pow(2, a.zoom - b.zoom),
                        t) * Math.pow(2, zoom_lerp - a.zoom),
                    interp(a.column,
                        b.column * Math.pow(2, a.zoom - b.zoom),
                        t) * Math.pow(2, zoom_lerp - a.zoom),
                    zoom_lerp);
            }
        };

        // The screen path means that the b
        // coordinate should maintain its point on screen
        // throughout the transition, but the map
        // should move to its zoom level
        paths.about = function(a, b, t, static_coord) {
            var zoom_lerp = interp(a.zoom, b.zoom, t);

            // center x, center y
            var cx = map.dimensions.x / 2,
                cy = map.dimensions.y / 2,
                // tilesize
                tx = map.tileSize.x,
                ty = map.tileSize.y;

            var startx = cx + tx * ((b.column * Math.pow(2, a.zoom - b.zoom)) - a.column);
            var starty = cy + ty * ((b.row  * Math.pow(2, a.zoom - b.zoom)) - a.row);

            var endx = cx + tx * ((b.column * Math.pow(2, zoom_lerp - b.zoom)) -
                (a.column * Math.pow(2, zoom_lerp - a.zoom)));
            var endy = cy + ty * ((b.row * Math.pow(2, zoom_lerp - b.zoom)) - (a.row *
                Math.pow(2, zoom_lerp - a.zoom)));

            if (static_coord) {
                static_coord.column = (a.column * Math.pow(2, zoom_lerp - a.zoom)) - ((startx - endx) / tx);
                static_coord.row = (a.row * Math.pow(2, zoom_lerp - a.zoom)) - ((starty - endy) / ty);
                static_coord.zoom = zoom_lerp;
            } else {
                return new MM.Coordinate(
                    (a.column * Math.pow(2, zoom_lerp - a.zoom)) - ((startx - endx) / tx),
                    (a.row * Math.pow(2, zoom_lerp - a.zoom)) - ((starty - endy) / ty),
                    zoom_lerp);
            }
        };

        var path = paths.screen;

        easey.t = function(t) {
            path(from, to, easing(t), static_coord);
            map.coordinate = static_coord;
            map.draw();
            return easey;
        };

        easey.future = function(parts) {
            var futures = [];
            for (var t = 0; t < parts; t++) {
                futures.push(path(from, to, t / (parts - 1)));
            }
            return futures;
        };

        var start;
        easey.resetRun = function () {
            start = (+ new Date());
            return easey;
        };

        easey.run = function(time, callback) {

            if (running) return easey.stop(function() {
                easey.run(time, callback);
            });

            if (!from) from = map.coordinate.copy();
            if (!to) to = map.coordinate.copy();
            time = time || 1000;
            start = (+new Date());
            running = true;

            function tick() {
                var delta = (+new Date()) - start;
                if (abort) {
                    abort = running = false;
                    abortCallback();
                    return (abortCallback = undefined);
                } else if (delta > time) {
                    if (to.zoom != from.zoom) map.dispatchCallback('zoomed', to.zoom - from.zoom);
                    running = false;
                    path(from, to, 1, static_coord);
                    map.coordinate = static_coord;
                    to = from = undefined;
                    map.draw();
                    if (callback) return callback(map);
                } else {
                    path(from, to, easing(delta / time), static_coord);
                    map.coordinate = static_coord;
                    map.draw();
                    MM.getFrame(tick);
                }
            }

            MM.getFrame(tick);
        };

        // Optimally smooth (constant perceived velocity) and
        // efficient (minimal path distance) zooming and panning.
        //
        // Based on "Smooth and efficient zooming and panning"
        // by Jarke J. van Wijk and Wim A.A. Nuij
        //
        // Model described in section 3, equations 1 through 5
        // Derived equation (9) of optimal path implemented below
        easey.optimal = function(V, rho, callback) {
            // Section 6 describes user testing of these tunable values
            V = V || 0.9;
            rho = rho || 1.42;

            function sqr(n) { return n*n; }
            function sinh(n) { return (Math.exp(n) - Math.exp(-n)) / 2; }
            function cosh(n) { return (Math.exp(n) + Math.exp(-n)) / 2; }
            function tanh(n) { return sinh(n) / cosh(n); }

            if (from) map.coordinate = from; // For when `from` not current coordinate
            else from = map.coordinate.copy();

            // Width is measured in coordinate units at zoom 0
            var TL = map.pointCoordinate(new MM.Point(0, 0)).zoomTo(0),
                BR = map.pointCoordinate(map.dimensions).zoomTo(0),
                w0 = Math.max(BR.column - TL.column, BR.row - TL.row),
                w1 = w0 * Math.pow(2, from.zoom - to.zoom),
                start = from.zoomTo(0),
                end = to.zoomTo(0),
                c0 = {x: start.column, y: start.row},
                c1 = {x: end.column, y: end.row},
                u0 = 0,
                u1 = Math.sqrt(sqr(c1.x - c0.x) + sqr(c1.y - c0.y));

            function b(i) {
                var n = sqr(w1) - sqr(w0) + (i ? -1: 1) * Math.pow(rho, 4) * sqr(u1 - u0),
                    d = 2 * (i ? w1 : w0) * sqr(rho) * (u1 - u0);
                return n/d;
            }

            function r(i) {
                return Math.log(-b(i) + Math.sqrt(sqr(b(i)) + 1));
            }

            var r0 = r(0),
                r1 = r(1),
                S = (r1 - r0) / rho;

            // Width
            var w = function(s) {
                return w0 * cosh(r0) / cosh (rho * s + r0);
            };

            // Zoom
            var u = function(s) {
                return (w0 / sqr(rho)) * cosh(r0) * tanh(rho * s + r0) - (w0 / sqr(rho)) * sinh(r0) + u0;
            };

            // Special case, when no panning necessary
            if (Math.abs(u1) < 0.000001) {
                if (Math.abs(w0 - w1) < 0.000001) return;

                // Based on section 4
                var k = w1 < w0 ? -1 : 1;
                S = Math.abs(Math.log(w1/w0)) / rho;
                u = function(s) {
                    return u0;
                };
                w = function(s) {
                    return w0 * Math.exp(k * rho * s);
                };
            }

            var oldpath = path;
            path = function (a, b, t, static_coord) {
                if (t == 1) {
                    if (static_coord) {
                        static_coord.row = to.row;
                        static_coord.column = to.column;
                        static_coord.zoom = to.zoom;
                    }
                    return to;
                }
                var s = t * S,
                    us = u(s),
                    z = a.zoom + (Math.log(w0/w(s)) / Math.LN2),
                    x = interp(c0.x, c1.x, us/u1 || 1),
                    y = interp(c0.y, c1.y, us/u1 || 1);

                var power = Math.pow(2, z);
                if (static_coord) {
                    static_coord.row = y * power;
                    static_coord.column = x * power;
                    static_coord.zoom = z;
                } else {
                    return new MM.Coordinate(y * power, x * power, z);
                }
            };

            easey.run(S / V * 1000, function(m) {
                path = oldpath;
                if (callback) callback(m);
            });
        };

        return easey;
    };

    this.easey = easey;
    if (typeof this.mapbox == 'undefined') this.mapbox = {};
    this.mapbox.ease = easey;
})(this, MM);


src/easey.handlers.js
;(function(context, MM) {

    var easey_handlers = {};

    easey_handlers.TouchHandler = function() {
        var handler = {},
            map,
            panner,
            maxTapTime = 250,
            maxTapDistance = 30,
            maxDoubleTapDelay = 350,
            locations = {},
            scale = 1,
            startDistance = 1,
            taps = [],
            wasPinching = false,
            lastPinchCenter = null,
            p0 = new MM.Point(0, 0),
            p1 = new MM.Point(0, 0);

        function focusMap(e) {
            map.parent.focus();
        }

        function clearLocations() {
            scale = 1;
            startDistance = 1;

            for (var loc in locations) {
                if (locations.hasOwnProperty(loc)) {
                    delete locations[loc];
                }
            }
        }

        function updateTouches (e) {
            var eventScale = (e.scale === undefined) ? scale : e.scale;

            for (var i = 0; i < e.touches.length; i += 1) {
                var t = e.touches[i];
                if (t.identifier in locations) {
                    var l = locations[t.identifier];
                    l.x = t.clientX;
                    l.y = t.clientY;
                    l.scale = eventScale;
                } else {
                    locations[t.identifier] = {
                        scale: eventScale,
                        startPos: { x: t.clientX, y: t.clientY },
                        startZoom: map.zoom(),
                        x: t.clientX,
                        y: t.clientY,
                        time: new Date().getTime()
                    };
                }
            }
        }

        function touchStartMachine(e) {
            if (!panner) panner = panning(map, 0.10);
            MM.addEvent(e.touches[0].target, 'touchmove',
                touchMoveMachine);
            MM.addEvent(e.touches[0].target, 'touchend',
                touchEndMachine);
            if (e.touches[1]) {
                MM.addEvent(e.touches[1].target, 'touchmove',
                    touchMoveMachine);
                MM.addEvent(e.touches[1].target, 'touchend',
                    touchEndMachine);

                if (e.scale === undefined) {
                    var _t0 = e.touches[0],
                        _t1 = e.touches[1],
                        _p0 = new MM.Point(0, 0),
                        _p1 = new MM.Point(0, 0);

                        _p0.x = _t0.clientX;
                        _p0.y = _t0.clientY;
                        _p1.x = _t1.clientX;
                        _p1.y = _t1.clientY;
                        startDistance = MM.Point.distance(_p0, _p1);
                }
            }
            updateTouches(e);
            panner.down(e.touches[0]);
            return MM.cancelEvent(e);
        }

        function touchMoveMachine(e) {
            switch (e.touches.length) {
                case 1:
                    panner.move(e.touches[0]);
                    break;
                case 2:
                    onPinching(e);
                    break;
            }
            updateTouches(e);
            return MM.cancelEvent(e);
        }

        // Handle a tap event - mainly watch for a doubleTap
        function onTap(tap) {
            if (taps.length &&
                (tap.time - taps[0].time) < maxDoubleTapDelay) {
                onDoubleTap(tap);
                taps = [];
                return;
            }
            taps = [tap];
        }

        // Handle a double tap by zooming in a single zoom level to a
        // round zoom.
        function onDoubleTap(tap) {
            // zoom in to a round number
            easey().map(map)
            .to(map.pointCoordinate(tap).zoomTo(map.getZoom() + 1))
            .path('about').run(200, function() {
                map.dispatchCallback('zoomed');
                clearLocations();
            });
        }

        function onPinching(e) {
            // use the first two touches and their previous positions
            var t0 = e.touches[0],
                t1 = e.touches[1];
            p0.x = t0.clientX;
            p0.y = t0.clientY;
            p1.x = t1.clientX;
            p1.y = t1.clientY;
            l0 = locations[t0.identifier];
            l1 = locations[t1.identifier];

            // mark these touches so they aren't used as taps/holds
            l0.wasPinch = true;
            l1.wasPinch = true;

            // scale about the center of these touches
            var center = MM.Point.interpolate(p0, p1, 0.5);

            // e.scale isn't available to Android or Windows 7 browsers
            var eventScale = (e.scale === undefined) ? scale = MM.Point.distance(p0, p1) / startDistance : e.scale;

            map.zoomByAbout(
                Math.log(eventScale) / Math.LN2 - Math.log(l0.scale) / Math.LN2,
                center);

            // pan from the previous center of these touches
            prevX = l0.x + (l1.x - l0.x) * 0.5;
            prevY = l0.y + (l1.y - l0.y) * 0.5;
            map.panBy(center.x - prevX,
                      center.y - prevY);
            wasPinching = true;
            lastPinchCenter = center;
        }

        // When a pinch event ends, round the zoom of the map.
        function onPinched(touch) {
            var z = map.getZoom(), // current zoom
                tz = locations[touch.identifier].startZoom > z ? Math.floor(z) : Math.ceil(z);

            easey().map(map).point(lastPinchCenter).zoom(tz)
                .path('about').run(300);
            clearLocations();
            wasPinching = false;
        }

        function touchEndMachine(e) {
            MM.removeEvent(e.target, 'touchmove',
                touchMoveMachine);
            MM.removeEvent(e.target, 'touchend',
                touchEndMachine);
            var now = new Date().getTime();

            // round zoom if we're done pinching
            if (e.touches.length <= 1 && wasPinching) {
                onPinched(e.changedTouches[0]);
            }

            panner.up();

            // Look at each changed touch in turn.
            for (var i = 0; i < e.changedTouches.length; i += 1) {
                var t = e.changedTouches[i],
                loc = locations[t.identifier];
                // if we didn't see this one (bug?)
                // or if it was consumed by pinching already
                // just skip to the next one
                if (!loc || loc.wasPinch) {
                    continue;
                }

                // we now know we have an event object and a
                // matching touch that's just ended. Let's see
                // what kind of event it is based on how long it
                // lasted and how far it moved.
                var pos = { x: t.clientX, y: t.clientY },
                time = now - loc.time,
                travel = MM.Point.distance(pos, loc.startPos);
                if (travel > maxTapDistance) {
                    // we will to assume that the drag has been handled separately
                } else if (time > maxTapTime) {
                    // close in space, but not in time: a hold
                    pos.end = now;
                    pos.duration = time;
                } else {
                    // close in both time and space: a tap
                    pos.time = now;
                    onTap(pos);
                }
            }

            // Weird, sometimes an end event doesn't get thrown
            // for a touch that nevertheless has disappeared.
            // Still, this will eventually catch those ids:

            var validTouchIds = {};
            for (var j = 0; j < e.touches.length; j++) {
                validTouchIds[e.touches[j].identifier] = true;
            }
            for (var id in locations) {
                if (!(id in validTouchIds)) {
                    delete validTouchIds[id];
                }
            }

            return MM.cancelEvent(e);
        }

        handler.init = function(x) {
            map = x;

            MM.addEvent(map.parent, 'touchstart',
                touchStartMachine);
        };

        handler.remove = function() {
            if (!panner) return;
            MM.removeEvent(map.parent, 'touchstart',
                touchStartMachine);
            panner.remove();
        };

        return handler;
    };

    easey_handlers.DoubleClickHandler = function() {
        var handler = {},
            map;

        function doubleClick(e) {
            // Ensure that this handler is attached once.
            // Get the point on the map that was double-clicked
            var point = MM.getMousePoint(e, map);
            z = map.getZoom() + (e.shiftKey ? -1 : 1);
            // use shift-double-click to zoom out
            easey().map(map)
                .to(map.pointCoordinate(MM.getMousePoint(e, map)).zoomTo(z))
                .path('about').run(100, function() {
                map.dispatchCallback('zoomed');
            });
            return MM.cancelEvent(e);
        }

        handler.init = function(x) {
            map = x;
            MM.addEvent(map.parent, 'dblclick', doubleClick);
            return handler;
        };

        handler.remove = function() {
            MM.removeEvent(map.parent, 'dblclick', doubleClick);
        };

        return handler;
    };

    easey_handlers.MouseWheelHandler = function() {
        var handler = {},
            map,
            _zoomDiv,
            ea = easey(),
            prevTime,
            precise = false;

        function mouseWheel(e) {
            var delta = 0;
            prevTime = prevTime || new Date().getTime();

            try {
                _zoomDiv.scrollTop = 1000;
                _zoomDiv.dispatchEvent(e);
                delta = 1000 - _zoomDiv.scrollTop;
            } catch (error) {
                delta = e.wheelDelta || (-e.detail * 5);
            }

            // limit mousewheeling to once every 200ms
            var timeSince = new Date().getTime() - prevTime;

            function dispatchZoomed() {
                map.dispatchCallback('zoomed');
            }

            if (!ea.running()) {
              var point = MM.getMousePoint(e, map),
                  z = map.getZoom();
              ea.map(map)
                .easing('easeOut')
                .to(map.pointCoordinate(MM.getMousePoint(e, map)).zoomTo(z + (delta > 0 ? 1 : -1)))
                .path('about').run(100, dispatchZoomed);
                prevTime = new Date().getTime();
            } else if (timeSince > 150){
                ea.zoom(ea.to().zoom + (delta > 0 ? 1 : -1)).from(map.coordinate).resetRun();
                prevTime = new Date().getTime();
            }

            // Cancel the event so that the page doesn't scroll
            return MM.cancelEvent(e);
        }

        handler.init = function(x) {
            map = x;
            _zoomDiv = document.body.appendChild(document.createElement('div'));
            _zoomDiv.style.cssText = 'visibility:hidden;top:0;height:0;width:0;overflow-y:scroll';
            var innerDiv = _zoomDiv.appendChild(document.createElement('div'));
            innerDiv.style.height = '2000px';
            MM.addEvent(map.parent, 'mousewheel', mouseWheel);
            return handler;
        };

        handler.precise = function(x) {
            if (!arguments.length) return precise;
            precise = x;
            return handler;
        };

        handler.remove = function() {
            MM.removeEvent(map.parent, 'mousewheel', mouseWheel);
            _zoomDiv.parentNode.removeChild(_zoomDiv);
        };

        return handler;
    };

    easey_handlers.DragHandler = function() {
        var handler = {},
            map,
            panner;

        function focusMap(e) {
            map.parent.focus();
        }

        function mouseDown(e) {
            if (e.shiftKey || e.button == 2) return;
            MM.addEvent(document, 'mousemove', mouseMove);
            MM.addEvent(document, 'mouseup', mouseUp);
            panner.down(e);
            map.parent.style.cursor = 'move';
            return MM.cancelEvent(e);
        }

        function mouseMove(e) {
            panner.move(e);
            return MM.cancelEvent(e);
        }

        function mouseUp(e) {
            MM.removeEvent(document, 'mousemove', mouseMove);
            MM.removeEvent(document, 'mouseup', mouseUp);
            panner.up();
            map.parent.style.cursor = '';
            return MM.cancelEvent(e);
        }

        handler.init = function(x) {
            map = x;
            MM.addEvent(map.parent, 'click', focusMap);
            MM.addEvent(map.parent, 'mousedown', mouseDown);
            panner = panning(map);
        };

        handler.remove = function() {
            MM.removeEvent(map.parent, 'click', focusMap);
            MM.removeEvent(map.parent, 'mousedown', mouseDown);
            panner.up();
            panner.remove();
        };

        return handler;
    };


    function panning(map, drag) {

        var p = {};
        drag = drag || 0.15;

        var speed = { x: 0, y: 0 },
            dir = { x: 0, y: 0 },
            removed = false,
            nowPoint = null,
            oldPoint = null,
            moveTime = null,
            prevMoveTime = null,
            animatedLastPoint = true,
            t,
            prevT = new Date().getTime();

        p.down = function(e) {
            nowPoint = oldPoint = MM.getMousePoint(e, map);
            moveTime = prevMoveTime = +new Date();
        };

        p.move = function(e) {
            if (nowPoint) {
                if (animatedLastPoint) {
                    oldPoint = nowPoint;
                    prevMoveTime = moveTime;
                    animatedLastPoint = false;
                }
                nowPoint = MM.getMousePoint(e, map);
                moveTime = +new Date();
            }
        };

        p.up = function() {
            if (+new Date() - prevMoveTime < 50) {
                dt = Math.max(1, moveTime - prevMoveTime);
                dir.x = nowPoint.x - oldPoint.x;
                dir.y = nowPoint.y - oldPoint.y;
                speed.x = dir.x / dt;
                speed.y = dir.y / dt;
            } else {
                speed.x = 0;
                speed.y = 0;
            }
            nowPoint = oldPoint = null;
            moveTime = null;
        };

        p.remove = function() {
            removed = true;
        };

        function animate(t) {
            var dt = Math.max(1, t - prevT);
            if (nowPoint && oldPoint) {
                if (!animatedLastPoint) {
                    dir.x = nowPoint.x - oldPoint.x;
                    dir.y = nowPoint.y - oldPoint.y;
                    map.panBy(dir.x, dir.y);
                    animatedLastPoint = true;
                }
            } else {
                // Rough time based animation accuracy
                // using a linear approximation approach
                speed.x *= Math.pow(1 - drag, dt * 60 / 1000);
                speed.y *= Math.pow(1 - drag, dt * 60 / 1000);
                if (Math.abs(speed.x) < 0.001) {
                    speed.x = 0;
                }
                if (Math.abs(speed.y) < 0.001) {
                    speed.y = 0;
                }
                if (speed.x || speed.y) {
                    map.panBy(speed.x * dt, speed.y * dt);
                }
            }
            prevT = t;
            if (!removed) MM.getFrame(animate);
        }

        MM.getFrame(animate);
        return p;
    }


    this.easey_handlers = easey_handlers;

})(this, MM);
